mtt/src/mtt.rs

```diff
@@ -0,0 +1,1190 @@
+            let region_memory_type = mtt_range.get_region_memory_type(addr, region_size)?;
+            if region_memory_type != memory_type {
+                return Ok(false);
+            }
+        }
+        Ok(true)
+    }
+
+    fn set_range_type(
+        addr: SupervisorPageAddr,
+        len: usize,
+        memory_type: MttMemoryType,
+        get_l1_page: &mut dyn FnMut() -> Option<SupervisorPageAddr>,
+    ) -> Result<bool> {
+        let mtt = Mtt::get()?;
+        let l2_dir = mtt.inner.lock();
+        let mut mtt_range = MttRange::new(addr, len, l2_dir)?;
+        let mut mtt_needs_invalidation = false;
+        while let Some((addr, region_size)) = mtt_range.next() {
+            if mtt_range.set_region_memory_type(addr, region_size, memory_type, get_l1_page)? {
+                mtt_needs_invalidation = true;
+            }
+        }
+        Ok(mtt_needs_invalidation)
+    }
+
+    /// Marks the region of length `len` starting with `addr` as confidential in
+    /// the Memory Tracking Table (MTT). On success, the function returns whether
+    /// the caller should make an ECALL to invalidate the MTT.
+    /// `len` must be a multiple of a 4K-page size. The region cannot exceed the
+    /// maximum physical page address (currently defined as 46-bits of PA).
+    /// # Safety
+    /// `get_l1_page` will be called to allocate a 4K-page if the function determines
+    /// that the region mapping requires a MTT mapping at a 4K granularity.
+    /// The callback must ensure that the allocated page isn't aliased and remains allocated
+    /// until it's explictly released (TBD), and it must not perform any MTT related
+    /// operations as it will result in a deadlock.
+    pub unsafe fn set_confidential(
+        addr: SupervisorPageAddr,
+        len: usize,
+        get_l1_page: &mut dyn FnMut() -> Option<SupervisorPageAddr>,
+    ) -> Result<bool> {
+        Mtt::set_range_type(addr, len, MttMemoryType::Confidential, get_l1_page)
+    }
+
+    /// Marks the region of length `len` starting with `addr` as non-confidential in
+    /// the Memory Tracking Table (MTT). On success, the function returns whether
+    /// the caller should make an ECALL to invalidate the MTT.
+    /// `len` must be a multiple of a 4K-page size. The region cannot exceed the
+    /// maximum physical page address (currently defined as 46-bits of PA).
+    /// # Safety
+    /// `get_l1_page` will be called to allocate a 4K-page if the function determines
+    /// that the region mapping requires a MTT mapping at a 4K granularity.
+    /// The callback must ensure that the allocated page isn't aliased and remains allocated
+    /// until it's explictly released (TBD), and it must not perform any MTT related
+    /// operations as it will result in a deadlock.
+    pub unsafe fn set_non_confidential(
+        addr: SupervisorPageAddr,
+        len: usize,
+        get_l1_page: &mut dyn FnMut() -> Option<SupervisorPageAddr>,
+    ) -> Result<bool> {
```

Comment:
(copying your previous comment because github sucks and won't let me continue that thread in this review)

> Please see the updates.
>
> 1. set_confidential and set_non_confidential have been changed to unsafe to allow expression of the constraints on the call back
> 2. l1_slice_from_pfn is unsafe as well.

why did we go back to a closure returning a raw address (`SupervisorPageAddr`) rather than a owning reference to the memory (`Page<InternalClean>`)? the latter is safe, at least in terms of the public API. see the usage of `get_pte_page` in `PageTable`. of course internally you may need to call an `unsafe` function with the raw address of the page -- like `l1_slice_from_pfn` -- but that's fine, we do similar things in `PageTable`.

to summarize my point about safety of `Page` vs raw address:

```
fn takes_a_page(page: Page<InternalClean>) {
    // Safety: we uniquely own the memory referred to by `page` and it is valid for reads and writes
    unsafe { write_to_pointer(page.addr()) }:
}

// Safety: `addr` must be valid for writes
unsafe fn writes_to_pointer(addr: SupervisorPageAddr) {
    core::ptr::write(...);
}
```

BUT: i do think these functions are unsafe, just not for the reason stated. the unsafety is that we're effectively destroying the contents of the memory being converted, and the `SupervisorPageAddr` type says nothing about whether or not that is safe to do.

Reply:
Changed.